### ISSUE TITLE
Read/Write permission issue

### ISSUE THREAD

**Issue by pdhimal1** [bitnami/bitnami-docker-phabricator#77]
This problem was supposed to be fixed by now.

I still get the following permision issue in bitnami/phabricator:2017 (which i think is the latest one?)
```
phab-container  | [07-Sep-2017 19:12:23 UTC] [2017-09-07 19:12:23] EXCEPTION: (CommandException) Command failed with error #1!
phab-container | COMMAND
phab-container | mkdir -p '/opt/bitnami/phabricator/data/d7/a6'
phab-container | 
phab-container | STDOUT
phab-container | (empty)
phab-container | 
phab-container | STDERR
phab-container | mkdir: cannot create directory '/opt/bitnami/phabricator/data/d7': Permission denied
```

Here's my `ls -l` result from `/bitnami/phabricator`
```
root@55b1b65f9ebf:/bitnami/phabricator# ls -l
total 0
drwxr-xr-x. 5 root        root        67 Sep  7 18:58 conf
drwxrwxr-x. 8 daemon      daemon      60 Sep  7 18:59 data
drwxr-xr-x. 4 phabricator phabricator 22 Sep  7 19:11 repo
```

Do I need to change permissions in `/opt/bitnami/phabricator` too?

Which user is supposed to own `/bitnami/phabricator/data`?

Its mentioned on some issues that the `phabricator` user is supposed to own `/bitnami/phabricator`.

However,  `chown -R phabricator:phabricator /bitmami/phabricator` results in local disk space not readable/writable error in the phabricator UI as a setup issue.

**Comment by pdhimal1** [comment]
Here's my output from docker info 
```
Server Version: 17.06.0-ce
Storage Driver: devicemapper
 Pool Name: docker-253:1-50362147-pool
 Pool Blocksize: 65.54kB
 Base Device Size: 10.74GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 11.73MB
 Data Space Total: 107.4GB
 Data Space Available: 12.07GB
 Metadata Space Used: 581.6kB
 Metadata Space Total: 2.147GB
 Metadata Space Available: 2.147GB
 Thin Pool Minimum Free Space: 10.74GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.26.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.703GiB
Name: atesthub2.enews.nrl.navy.mil
ID: HTC5:52G4:AF6T:NFNE:DUXN:QNNP:HAN4:VOK5:MNFZ:NSOL:SQTB:ZJOQ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: devicemapper: usage of loopback devices is strongly discouraged for production use.
         Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.

```

**Comment by pdhimal1** [comment]
Any updates on this?
 
Phabricator mentions that the webserver user should have read/write permission in the data directory.
What is the webserver user in a bitnami container? Is it the `www-data` user or the `daemon` user?

Why do you guys recommand giving permission to the `phabricator` user?

**Comment by chipironcin** [comment]
Hi @pdhimal1 
The default webserver user in a Bitnami container is `daemon`.
Folders in `/bitnami/phabricator` is just where the original data is backed up. Links to those folders can be found in the main Phabricator folder at `/opt/bitnami/phabricator`.

If you execute `chown -R phabricator:phabricator /bitmami/phabricator` you are not allowing the webserver to write in `data` folder.

What command are you trying to execute?
You can easily fix this situation by allowing the `phabricator` user to write the `data` folder.

    chown -R phabricator:daemon /bitnami/phabricator/data
    chmod 775 /bitnami/phabricator/data

Then, retry your command.
We will evaluate this fix and include it in the next release of Bitnami Phabricator container if it applies.
Thanks for reporting

**Comment by pdhimal1** [comment]
Doesn't work

Here's my `ls -l` result from `/bitnami/phabricator`

```
root@3c256d12ccc2:/bitnami/phabricator# ls -l
total 4
drwxr-xr-x.   5 root        root          67 Sep 15 15:40 conf
drwxrwxr-x. 109 phabricator daemon      4096 Sep 15 15:49 data
drwxr-xr-x.   3 phabricator phabricator   14 Sep 15 15:48 repo
```

If `"The default webserver user in a Bitnami container is daemon."`, isn't daemon supposed to be the owner of the `/bitnami/phabricator/data` folder?

**Comment by pdhimal1** [comment]
This happens when trying to import a repo from github

```
phab-container |[15-Sep-2017 17:28:06 UTC]   #23 PhabricatorTaskmasterDaemon::run() called at [<phutil>/src/daemon/PhutilDaemon.php:219]
phab-container | [15-Sep-2017 17:28:06 UTC]   #24 PhutilDaemon::execute() called at [<phutil>/scripts/daemon/exec/exec_daemon.php:131]
phab-container | [15-Sep-2017 17:28:07 UTC] [2017-09-15 17:28:07] EXCEPTION: (CommandException) Command failed with error #1!
phab-container | COMMAND
phab-container | mkdir -p '/opt/bitnami/phabricator/data/ab/96'
phab-container | 
phab-container | STDOUT
phab-container | (empty)
phab-container | 
phab-container | STDERR
phab-container | mkdir: cannot create directory '/opt/bitnami/phabricator/data/ab': Permission denied
phab-container |  at [<phutil>/src/future/exec/ExecFuture.php:369]
phab-container | [15-Sep-2017 17:28:07 UTC] arcanist(head=stable, ref.stable=92f7b2986cd7), phabricator(head=stable, ref.stable=b9669edbbd41), phutil(head=stable, ref.stable=f360228427fb)

```
And many more error messages like this. However it looks like the repo is fully imported.

Looking into the data folder, here's my premissions:
```
root@d515bd5967c1:/bitnami/phabricator/data# ls -la
total 0
drwxrwxr-x. 9 daemon daemon 69 Sep 15 17:28 .
drwxr-xr-x. 5 root   root   58 Sep 15 17:21 ..
drwxr-xr-x. 3 daemon daemon 15 Sep 15 17:24 60
drwxr-xr-x. 3 daemon daemon 15 Sep 15 17:24 b1
drwxr-xr-x. 3 daemon daemon 15 Sep 15 17:24 b7
drwxr-xr-x. 3 daemon daemon 15 Sep 15 17:28 c0
drwxr-xr-x. 3 daemon daemon 15 Sep 15 17:21 da
drwxr-xr-x. 3 daemon daemon 15 Sep 15 17:24 dd
drwxr-xr-x. 3 daemon daemon 15 Sep 15 17:24 e9
```

Looks like the ones mentioned in the error message are not being created. I see some files in this data folder, which tells me that phabricator has the correct permission. 

I hope this helps you trying to get this bug fixed.

**Comment by chipironcin** [comment]
>If "The default webserver user in a Bitnami container is daemon.", isn't daemon supposed to be the owner of the /bitnami/phabricator/data folder?
By default it is, but it seems is the user `phabricator` the one that tries to write in there so I suggested giving permission to both users.

I am able to reproduce the issues, though the repo is fully imported successfully.
It seems both users (daemon and phabricator) try to write to data folder.
We can give them permissions to write in the data folder, but they will also try to write under folders in that data folder.

The issue is that the user phabricator is trying to create a directory inside a folder owned by daemon in the data directory. And it fails.

I cannot find a suitable fix for this. I think this is an issue related to the application rather than the container.

If you are interested in removing that error from the logs, despite the application working well, I suggest you to contact the team developing Phabricator.

**Comment by pdhimal1** [comment]
I was able to fix the permission issue by using ACL on my container
```
apt-get -y install acl  
setfacl -Rm d:u:phabricator:rwX,u:phabricator:rwX /bitnami/phabricator/data/
```

There are no permission related warning in my container.

However I see this new error:

**Comment by chipironcin** [comment]
I haven't seen it before but it may be due to the previous failures while writing in the data folder.
`PHID-FILE-vqt7a7eihkybj64xysxl` makes reference to an uploaded file.

Personally I have experienced some temporal errors similar to this one in the Phabricator UI (not using this container). They were gone after retying the action.

If you keep experiencing them, please open a new issue.
I will close this one as there is no possible applicable workaround for this image right now.
Thank you for posting back your workaround 👍

**Comment by felixfbecker** [comment]
I am also seeing this with the latest version of the image, and it seems to make it impossible to import repos from GitHub. As this issue was closed, is there no way to fix this?

**Comment by juan131** [comment]
Hi @felixfbecker 

Could you please provide the steps you followed to reproduce the issue?

**Comment by rabidsloth** [comment]
I just ran into this issue as well. I'm deploying `bitnami/phabricator:latest` to kubernetes with a mounted volume at `/bitnami/phabricator`. Any time the container gets restarted, everything within the data directory gets `daemon:daemon` and I have to manually `chown -R phabricator:phabricator` that directory to get the app to start throwing errors about permission issues in there.

**Comment by miguelaeh** [comment]
Hi @rabidsloth ,
It seems to be a problem with the initialization logic, the permissions are configure correctly for the `/bitnami/phabricator/repo` dir but not for `/bitnami/phabricator/data`.
Thank you very much for reporting it, we will fix it as soon as possible.

**Comment by tinder-maxwellelliott** [comment]
As of `bitnami/phabricator:latest` this is still happening as of today. We had to chown `opt/bitnami/phabricator/data` so that we could actually close diffs and have them linked back to our revisions.

If you are seeing an issue where you are landing revisions but Phab is not closing them, running `chown a+w -R opt/bitnami/phabricator/data` will fix it (not a production fix but it will unblock you)

**Comment by tinder-maxwellelliott** [comment]
> As of `bitnami/phabricator:latest` this is still happening as of today. We had to chown `opt/bitnami/phabricator/data` so that we could actually close diffs and have the commits linked back to our revisions.
> 
> If you are seeing an issue where you are landing revisions but Phab is not closing them, running `chown a+w -R opt/bitnami/phabricator/data` will fix it (not a production fix but it will unblock you)

This approach still does not work since the chown will not pickup new subdirectories that Phab creates. In the end we ended up moving from local volumes to MySQL and it worked. This approach is documented here https://github.com/bitnami/bitnami-docker-phabricator/issues/70

**Comment by marcosbc** [comment]
Hi @tinder-maxwellelliott, could you confirm if the volume has data thad has been created/modified/removed from outside of the container?

Also with this we're bit confused:

> In the end we ended up moving from local volumes to MySQL and it worked.

Did you mean from local volumes to Docker volumes?

**Comment by tinder-maxwellelliott** [comment]
We are still using this volume setup in our docker-compose

```
    volumes:
      - 'phabricator_data:/bitnami'
```

But we had to run the following commands to make Phab behave and have the correct permissions

```
bin/config set storage.mysql-engine.max-size 8388608"
/bin/files migrate --engine blob --all"
/bin/config set storage.local-disk.path ''"
```

**Comment by marcosbc** [comment]
You seem to be right, the Phabricator daemon services are running as the `phabricator` user, while the Phabricator PHP UI is running as the `daemon` user, and the `data` folder by default belongs to `daemon`.

We're glad you could workaround this issue for now by using MySQL as a storage engine. 

For now I have created an internal task for unifying all Phabricator services into a single user to avoid issues like these. Once we start working on the task we will add any update.

**Comment by juan131** [comment]
Hi everyone

I just want to let you know that we just released a new major version including several improvements that break backwards compatibility. Please read carefully the Notable Section below:

- https://github.com/bitnami/bitnami-docker-phabricator/blob/master/README.md#notable-changes

This new image runs, by default, the container with a non-root user (`1001` although it can be replaced by any other). As a consequence, the `daemon` and `phabricator` users are no longer used to launch the services.
Running this image as a non-root user, you should not face any permissions issues anymore.

In case you want to continue using the image as a root container, the `daemon` and `phabricator` users will be used and, the default permissions/ownership under the "data" and "repo" directories are the ones below:

- `/bitnami/phabricator/data/`:
  - Ownership: `daemon:daemon`
  - Directores `755`
  - Files `644`
- `/bitnami/phabricator/var/repo/`:
  - Ownership: `phabricator:phabricator`
  - Directores `755`
  - Files `644`
